### PR TITLE
vulkan: Clamp VK_EXT_memory_budget heapBudget/heapUsage to guest heap sizes

### PR DESCRIPTION
**Pull request by utzcoz**
VK_EXT_memory_budget is passed through to the host driver, so heapBudget and heapUsage are reported against the host's heap sizes. gfxstream clamps oversized guest heap sizes (kMaxSafeHeapSize), which let the guest observe heapBudget/heapUsage larger than the heap size it sees, violating the spec invariant heapBudget <= VkMemoryHeap::size.

Add EmulatedPhysicalDeviceMemoryProperties::clampMemoryBudgetToGuestHeapSizes() and call it from on_vkGetPhysicalDeviceMemoryProperties2 to clamp each per-heap value to the guest-visible heap size. Register the VkPhysicalDeviceMemoryProperties2 and VkPhysicalDeviceMemoryBudgetPropertiesEXT struct ids so vk_find_struct can locate the budget struct in the pNext chain. Add unit tests covering the clamp, the within-bounds pass-through case, and the null (unchained) case.

See https://docs.vulkan.org/refpages/latest/refpages/source/VkPhysicalDeviceMemoryBudgetPropertiesEXT.html#_description

### CODE REVIEW

**Comment by utzcoz** [comment]
Hi @gurchetansingh @jmacnak , PTAL. Thanks.

File: host/vulkan/vk_emulated_physical_device_memory_tests.cpp [hunk 1]
```diff
@@ -477,6 +477,79 @@ TEST(VkGuestMemoryUtilsTest, VulkanAMDCoherentFlagsNotLeakedToGuest) {
     }
 }
 
+TEST(VkGuestMemoryUtilsTest, MemoryBudgetClampedToClampedGuestHeapSize) {
+    constexpr VkDeviceSize kMaxSafeHeapSize = 2ULL * 1024ULL * 1024ULL * 1024ULL;
```
Comment (jmacnak) [inline host/vulkan/vk_emulated_physical_device_memory_tests.cpp]: To avoid duplicating constants / making the test rely on impl behavior, could you:

* update `EmulatedPhysicalDeviceMemoryProperties` to have a `mMaxSafeHeapSize` member
* move the existing
  ```
  // Hide any bogus heap sizes from bad drivers with a reasonable default that will not
  // break the bank on 32-bit userspaces.
  static constexpr VkDeviceSize kMaxSafeHeapSize = 2ULL * 1024ULL * 1024ULL * 1024ULL;
  ```
  from `vk_emulated_physical_device_memory.cpp` to the `vk_emulated_physical_device_memory.h` with a rename as `kDefaultMaxSafeHeapSize`
* update the constructor to take a `VkDeviceSize maxSafeHeapSize` as a paremter defaulting to `kDefaultMaxSafeHeapSize`
* update the test to either explicitly use `kDefaultMaxSafeHeapSize` or some other test constant

File: host/vulkan/vk_emulated_physical_device_memory_tests.cpp [hunk 2]
```diff
@@ -477,6 +477,79 @@ TEST(VkGuestMemoryUtilsTest, VulkanAMDCoherentFlagsNotLeakedToGuest) {
...
+    EXPECT_EQ(budget.heapBudget[0], kMaxSafeHeapSize);
+    EXPECT_EQ(budget.heapUsage[0], kMaxSafeHeapSize);
+    EXPECT_EQ(budget.heapBudget[1], 768ULL * 1024 * 1024);
+    EXPECT_EQ(budget.heapUsage[1], 256ULL * 1024 * 1024);
+}
+
+TEST(VkGuestMemoryUtilsTest, MemoryBudgetWithinHeapSizesUnchanged) {
+    VkPhysicalDeviceMemoryProperties hostMemoryProperties = {};
+    hostMemoryProperties.memoryHeapCount = 1;
+    hostMemoryProperties.memoryHeaps[0] = {.size = 0x10000000, .flags = 0};
```
Comment (jmacnak) [inline host/vulkan/vk_emulated_physical_device_memory_tests.cpp]: nit: Could you use the more human readable `256ULL * 1024 * 1024` style in `MemoryBudgetWithinHeapSizesUnchanged` and `MemoryBudgetNullIsNoOp` and to match  `MemoryBudgetClampedToClampedGuestHeapSize` test? (I realize the existing above tests above use the 0x style but we should make those more human readabile as well)

**Review by jmacnak — CHANGES_REQUESTED** [review]
Looks good, 2 minor changes requested

**Comment by utzcoz** [comment]
Thanks for reviewing, I will update it to address comments today.

File: host/vulkan/vk_emulated_physical_device_memory_tests.cpp (hunk 1 above)
Comment (utzcoz) [reply host/vulkan/vk_emulated_physical_device_memory_tests.cpp]: Applied.

File: host/vulkan/vk_emulated_physical_device_memory_tests.cpp (hunk 2 above)
Comment (utzcoz) [reply host/vulkan/vk_emulated_physical_device_memory_tests.cpp]: Applied.

**Comment by utzcoz** [comment]
@jmacnak PTAL. Thanks.

**Review by jmacnak — APPROVED** [review]
LGTM, thank you for the change!
